rfcs/convergence/event-handlers-arguments.md

```diff
@@ -0,0 +1,164 @@
+
+## Background
+
+We should provide guidance on how events handlers (`onChange`, `onOpen`, etc.) are exposed across components. Today we have a discrepancy between `@fluentui/react-northstar` (Fluent UI Northstar) and `@fluentui/react` (Fluent UI v8) components:
+
+```tsx
+// v8
+// "newValue" is the new value of the component's main data, depending on the component type
+onChange: (ev: React.FormEvent, newValue:  ___);
+```
+
+```tsx
+// Northstar
+// "data" is props, with the new value mixed in on top
+// const data = { ...props, value: newValue }
+onChange: (ev: React.FormEvent, data: TProps);
+```
+
+For converged components (Fluent UI vNext) it's implemented in different ways 😪
+
+## Use cases for event handlers
+
+1. The primary data has changed, and I need the new value to support controlled component cases or to forward the data to a state management system.
+
+2. When the data changes I not only need the new value, but I need the original props of the component, so that I can differentiate which instance fired the change event to a general handler.
+
+There are a few types of components which might fire an `onChange` event:
+
+- Single literal value components (`value = "foo"`): `Input` and `Toggle`
+- Single-object value components (`value = { name: 'foo-item' }`): `RadioGroup`, `Select`
+- Multi-object selection components (`value = ["foo", "bar"]`): multiselect `Dropdown`, `DetailsList`, `Tree`
+
+## Problem statement
+
+We need to standardize event handlers across components and ensure that API is consistent to avoid confusion for customers. To clarify, we should avoid this case:
+
+```tsx
+// ❌ Each component in this example has completely different signature for event handlers
+<>
+  <Menu onOpen={(ev: React.ClickEvent, newOpen: boolean) => {}} />
+  <Input onOpen={(ev: React.ClickEvent, data: { value: string }) => {}} />
+</>
+```
+
+## Prior considerations
+
+Semantic UI React's thread for `onChange` which provided some background on thinking for Option B below:
+https://github.com/Semantic-Org/Semantic-UI-React/issues/623#issuecomment-261018287
+
+Material UI follows a similar approach to v8 code: the second argument represents the new value being changed. See [Select API](https://material-ui.com/api/select/) as an example of this. However the components aren't consistent - See [Input API](https://material-ui.com/api/input/) or [Switch API](https://material-ui.com/api/switch/) where neither have a second argument.
+
+## Detailed Design or Proposal
+
+### Option A (recommended):
+
+- Standardize on a `data` object where we provide the following minimum props:
+
+| Prop    | Description                                      |
+| ------- | ------------------------------------------------ |
+| `value` | new value for the data of the component          |
+| `props` | the original props given to the parent component |
+
+Example:
+
+```tsx
+interface ChangeData<TValue, TProps> {
+  value: TValue;
+  props: TProps;
```

Comment:
I don't see why parent props are necessary here, this is user input and we do not mutate it so why should we return all of this ? 

As you mention later it will also cause some 'smoke' and make the valuable properties harder to find.

The DTO is a 👍 but we should only return data that is pertinent to the change/mutation that is happening on the component for proper encapsulation. Otherwise if consumers decide to forward change handlers internally in their app then you break encapsulation of the component tree

Reply:
There are scenarios where the caller has multiple components hooked up to a single onChange callback. Without providing the original props, each instance would need to be closured with the identifiable data to handle them individually.

Also someone who looks at the event handler without reading the details might assume that `props.value` and `value` are equal.

One way to clear up some ambiguity might be to call it "originalProps" and pass along the original props. E.g. if `value` changes from 'a' to 'b', then `value` would be `b`, while `originalProps.value` would be 'a'.

Reply:
> each instance would need to be closured with the identifiable data to handle them individually

Isn't that a good thing ? in that case your multiple component are encapsulated into a single component and you should only pass on relevant data to its parent.